onedal/backend/data/data.cpp

```diff
@@ -0,0 +1,194 @@
+public:
+    numpy_deleter(PyArrayObject *a) : ndarray_(a) {}
+
+    void operator()(const void *ptr) {}
+
+    numpy_deleter &operator=(const numpy_deleter &) = delete;
+
+private:
+    PyArrayObject *ndarray_;
+};
+
+template <typename T, typename ConstDeleter>
+inline dal::homogen_table create_homogen_table(const T *data_pointer,
+                                               const std::size_t row_count,
+                                               const std::size_t column_count,
+                                               const dal::data_layout layout,
+                                               ConstDeleter &&data_deleter) {
+    return dal::homogen_table(data_pointer, row_count, column_count, data_deleter, layout);
+}
+
+template <typename T>
+inline dal::homogen_table make_homogen(PyArrayObject *array) {
+    size_t column_count = 1;
+
+    if (array_numdims(array) > 2) {
+        throw std::runtime_error("Input array has wrong dimensionality (must be 2d).");
+    }
+    T *data_pointer = reinterpret_cast<T *>(array_data(array));
+    const size_t row_count = static_cast<size_t>(array_size(array, 0));
+    if (array_numdims(array) == 2) {
+        column_count = static_cast<size_t>(array_size(array, 1));
+    }
+    const auto layout =
+        array_is_behaved_F(array) ? dal::data_layout::column_major : dal::data_layout::row_major;
+    auto res_table =
+        create_homogen_table(data_pointer, row_count, column_count, layout, numpy_deleter(array));
+    // we need it increment the ref-count if we use the input array in-place
+    // if we copied/converted it we already own our own reference
+    if (reinterpret_cast<PyArrayObject *>(data_pointer) == array)
+        Py_INCREF(array);
+    return res_table;
+}
+
+dal::table input_to_onedal_table(PyObject *obj) {
+    dal::table res;
+    if (obj == nullptr || obj == Py_None) {
+        return res;
+    }
+    if (is_array(obj)) {
+        PyArrayObject *ary = reinterpret_cast<PyArrayObject *>(obj);
+        if (array_is_behaved(ary) || array_is_behaved_F(ary)) {
+#define MAKE_HOMOGEN_TABLE(CType) res = make_homogen<CType>(ary);
+            SET_NPY_FEATURE(PyArray_DESCR(ary)->type,
+                            MAKE_HOMOGEN_TABLE,
+                            throw std::invalid_argument("Found unsupported array type"));
+#undef MAKE_HOMOGEN_TABLE
+        }
+        else {
+            throw std::invalid_argument(
+                "[input_to_onedal_table] Numpy input Could not convert Python object to onedal table.");
+        }
+    }
+    else {
+        throw std::invalid_argument(
+            "[input_to_onedal_table] Not avalible input format for convert Python object to onedal table.");
+    }
+    return res;
+}
+
+class vsp {
+public:
+    // we need a virtual destructor
+    virtual ~vsp(){};
+};
+
+// typed virtual shared pointer, for simplicity we make it a oneDAL shared pointer
+template <typename T>
+class tvsp : public vsp, public dal::array<T> {

```

Comment:
the name is confusing. why shared pointer? why it is inherited from `dal::array`?